rootfs/etc/nginx/lua/balancer/ewma.lua

```diff
@@ -125,16 +130,31 @@ local function shuffle_peers(peers, k)
   -- peers[1 .. k] will now contain a randomly selected k from #peers
 end
 
-local function pick_and_score(peers, k)
+local function pick_and_score(peers, k, chosen_endpoints)
   shuffle_peers(peers, k)
-  local lowest_score_index = 1
-  local lowest_score = score(peers[lowest_score_index])
-  for i = 2, k do
+  local lowest_score_index
+  local lowest_score
+  for i = 1, k do
+    local key = get_upstream_name(peers[i])
+    if chosen_endpoints[key] then
```

Comment:
Let's assume `peers` has 10 elements, out of which 2 are already tried. With the approach you implemented there's still a chance that we pick already tried upstream. But alternatively we can introduce a new list i.e `filtered_peers`, which is a copy of `peers` without the 2 already tried peers. That way we guarantee that we don't pick an already tried peer. I'm curious why you did not take this approach instead?

Reply:
Aha, I didn't notice there is a PICK_SET_SIZE limitation.
Just curiously, why we set a PICK_SET_SIZE limitation and its value is 2? The original implementation (https://github.com/twitter/finagle/blob/1bc837c4feafc0096e43c0e98516a8e1c50c4421/finagle-core/src/main/scala/com/twitter/finagle/loadbalancer/PeakEwma.scala) seems don't have the same limitation.

Reply:
That thing seems to be specifically about calculating peak ewma. I don't see shuffling or picking nodes in that code.

I think value 2 is based on "Michael Mitzenmacher. 2001. The Power of Two Choices in Randomized Load Balancing. IEEE Trans. Parallel Distrib. Syst. 12, 10 (October 2001), 1094-1104."